build.py

```diff
@@ -0,0 +1,228 @@
+
+'''
+GFX reconstruct build script
+'''
+
+import argparse
+import distutils.version
+import os
+import platform
+import re
+import shutil
+import subprocess
+import sys
+
+
+def is_windows():
+    '''
+    Check if the system is Windows
+    '''
+    return 'windows' == platform.system().lower()
+
+
+ARCHITECTURES = ['x64', 'x86']
+BUILD_ROOT = os.path.split(os.path.abspath(__file__))[0]
+BUILD_CONFIGS = {'debug': 'dbuild', 'release': 'build'}
+if is_windows():
+    BUILD_CONFIGS['debug'] = 'build'
+CMAKE_VERSION_3_13 = distutils.version.StrictVersion('3.13.0')
+CONFIGURATIONS = ['release', 'debug']
+VERSION = distutils.version.StrictVersion('0.0.0')
+
+
+class BuildError(Exception):
+    '''
+    Raised on a build error
+    '''
+
+
+def parse_args():
+    '''
+    Parse command line arguments
+    '''
+    arg_parser = argparse.ArgumentParser(
+        description="gfxreconstruct build script",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    arg_parser.add_argument('--version', dest='version',
+                            action='version', version=str(VERSION))
+    arg_parser.add_argument(
+        '-a', '--arch', dest='architecture',
+        metavar='ARCH', action='store',
+        choices=ARCHITECTURES, default=ARCHITECTURES[0],
+        help='Build target architecture. Can be one of: {0}'.format(
+                ', '.join(ARCHITECTURES)))
+    arg_parser.add_argument(
+        '-c', '--config', dest='configuration',
+        metavar='CONFIG', action='store',
+        choices=CONFIGURATIONS, default=CONFIGURATIONS[0],
+        help='Build target configuration. Can be one of: {0}'.format(
+            ', '.join(CONFIGURATIONS)))
+    arg_parser.add_argument(
+        '--clean', dest='clean', action='store_true', default=False,
+        help='Clean the build targets')
+    arg_parser.add_argument(
+        '--clobber', dest='clobber', action='store_true', default=False,
+        help='Clean the build targets, and remove build and intermediate files')
+    arg_parser.add_argument(
+        '--skip-update-deps', dest='skip_update_deps',
+        action='store_true', default=False,
+        help='Skip updating external dependencies')
+    arg_parser.add_argument(
+        '--skip-code-style', dest='skip_code_style',
+        action='store_true', default=False,
+        help='Skip applying C++ code style before compiling')
+    arg_parser.add_argument(
+        '--static-analysis', dest='static_analysis',
+        action='store_true', default=False,
+        help='Run static analysis on the code')
+    return arg_parser.parse_args()
+
+
+def update_external_dependencies(args):
+    '''
+    Update git submodules
+    '''
+    if not args.skip_update_deps:
+        update_git_submodule_result = subprocess.run(
+            ['git', 'submodule', 'update', '--init'], cwd=BUILD_ROOT)
+        if 0 != update_git_submodule_result.returncode:
+            raise BuildError('failed to update git submodules')
```

Comment:
This is fine if everything works; but if something goes wrong, important information (the exact return code, stdout, and stderr) that would be necessary for debugging the problem is lost.  The debugging engineer would have to dissect the script to determine the commands to run, re-run the commands manually, and hope it fails in the same way.

I recommend instead:
```
update_git_submodule_result = subprocess.run(
        ['git', 'submodule', 'update', '--init', '--recursive'],
        cwd=BUILD_ROOT, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
if 0 != update_git_submodule_result.returncode:
        raise BuildError('failed to update git submodules: {}'.format(update_git_submodule_result))
```
which will include in the result object the return code, stdout, and stderr.  A similar technique should be used for all `subprocess.run()` calls.